### PR TITLE
fix: SQL Targets ignore collation when evaluating column data types

### LINKED ISSUE #1125: bug: column collations should be ignored during SQL type comparisons (SQL Server)
### Singer SDK Version

0.11.1

### Python Version

3.10

### Bug scope

Targets (data type handling, batching, SQL object generation, etc.)

### Operating System

Mac

### Description

SQL Server target: Initial load of a table with a varchar column works nicely, but on second write, the loader fails with a ValueError, after attempting to change column types. In the example, a PG source table with a VARCHAR(64) column:

`ValueError: Unable to merge sql types: VARCHAR(64) COLLATE "SQL_Latin1_General_CP1_CI_AS", VARCHAR(64)`

My interpretation is that the target initially ran a create table statement with a VARCHAR(64) column, SQL Server added the collation as a detail, but when inserting again the target interprets varchar with the added collation as a different column type than VARCHAR(64) and tries to alter it, but fails.

An ugly fix is to explicitly declare the collation on table creation: https://github.com/radbrt/target-mssql/blob/37544ad6306095b66bef9f01d35211f486b9bf77/target_mssql/sinks.py#L262

### Code

_No response_

### PR DESCRIPTION
**Pull request by BuzzCutNorman**
This is an attempt to fix #1125

----
:books: Documentation preview :books:: https://meltano-sdk--1385.org.readthedocs.build/en/1385/

### CODE REVIEW

File: singer_sdk/connectors/sql.py [hunk 1]
```diff
@@ -959,10 +959,18 @@ def _adapt_column_type(
         Raises:
             NotImplementedError: if altering columns is not supported.
         """
+        collation_removed = False
         current_type: sqlalchemy.types.TypeEngine = self._get_column_type(
             full_table_name, column_name
         )
 
+        # remove collation if present and save it
+        if hasattr(current_type, "collation"):
```
Comment (edgarrmondragon) [inline singer_sdk/connectors/sql.py:968]: Missing coverage here may point to a useful refactor that can be unit-tested:

Wdyt about moving this block to a helper function that takes a type object and returns the collation, or `None` if the collation is falsy:

```python
def remove_collation(column_type: TypeEngine) -> "Collation" | None:  # don't know the actual type of a collation
    if hasattr(current_type, "collation") and current_type.collation:
        current_type_collation = current_type.collation
        setattr(current_type, "collation", None)
        return current_type_collation
    return None

def update_collation(column_type: TypeEngine, collation: "Collation" | None) -> None:
    """Replace column collation if it has one."""
    if hasattr(column_type, "collation") and collation:
        setattr(column_type, "collation", collation)
```

Later we can check if the collation is not `None` to set it in the new type

```python
        current_type_collation = remove_collation(current_type)

        ...

        update_collation(compatible_sql_type, current_type_collation)
```

**Review by edgarrmondragon — CHANGES_REQUESTED** [review]

File: singer_sdk/connectors/sql.py (hunk 1 above)
Comment (BuzzCutNorman) [reply singer_sdk/connectors/sql.py:968]: I think this looks great. I will work on getting these two functions in place and tests for them.

Comment (edgarrmondragon) [reply singer_sdk/connectors/sql.py:968]: Thanks @BuzzCutNorman!

Comment (BuzzCutNorman) [reply singer_sdk/connectors/sql.py]: @edgarrmondragon when you have a moment, please review the changes I made.

File: tests/core/test_connector_sql.py [hunk 2]
```diff
@@ -91,3 +91,42 @@ def test_get_column_ddl(
...
         )
         assert statement == rendered_statement
+
+    def test_remove_collation_text_type(self, connector):
+        remove_collation = getattr(connector, "remove_collation")
+        test_collation = "SQL_Latin1_General_CP1_CI_AS"
+        current_type = sqlalchemy.types.Text(collation=test_collation)
+        current_type, current_type_collation = remove_collation(current_type)
+        # Check collation was set to None by the function
+        assert current_type.collation == None
```
Comment (edgarrmondragon) [inline tests/core/test_connector_sql.py:101]:
```suggestion
        assert current_type.collation is None
```

File: tests/core/test_connector_sql.py [hunk 3]
```diff
@@ -91,3 +91,42 @@ def test_get_column_ddl(
...
+        # Check that we get the same collation we put in back out
+        assert current_type_collation == test_collation
+
+    def test_remove_collation_non_text_type(self, connector):
+        remove_collation = getattr(connector, "remove_collation")
+        current_type = sqlalchemy.types.Integer()
+        current_type, current_type_collation = remove_collation(current_type)
+        # Check there is not a collation attribute
+        assert hasattr(current_type, "collation") == False
```
Comment (edgarrmondragon) [inline tests/core/test_connector_sql.py:110]:
```suggestion
        assert not hasattr(current_type, "collation")
```

File: tests/core/test_connector_sql.py [hunk 4]
```diff
@@ -91,3 +91,42 @@ def test_get_column_ddl(
...
+        # Check that we get the same type we put in
+        assert str(current_type) == "INTEGER"
+        # Check that this variable is missing
+        assert not current_type_collation
```
Comment (edgarrmondragon) [inline tests/core/test_connector_sql.py:114]:
```suggestion
        assert current_type_collation is None
```

File: tests/core/test_connector_sql.py [hunk 5]
```diff
@@ -91,3 +91,42 @@ def test_get_column_ddl(
...
+        # Check collation was set to the value we put in
+        assert compatible_type.collation == test_collation
+
+    def test_update_collation_non_text_type(self, connector):
+        update_collation = getattr(connector, "update_collation")
+        test_collation = "SQL_Latin1_General_CP1_CI_AS"
+        compatible_type = sqlalchemy.types.Integer()
+        compatible_type = update_collation(compatible_type, test_collation)
+        # Check there is not a collation attribute
+        assert hasattr(compatible_type, "collation") == False
```
Comment (edgarrmondragon) [inline tests/core/test_connector_sql.py:130]:
```suggestion
        assert not hasattr(compatible_type, "collation")
```

File: tests/core/test_connector_sql.py [hunk 6]
```diff
@@ -91,3 +91,42 @@ def test_get_column_ddl(
             )
         )
         assert statement == rendered_statement
+
+    def test_remove_collation_text_type(self, connector):
+        remove_collation = getattr(connector, "remove_collation")
```
Comment (edgarrmondragon) [inline tests/core/test_connector_sql.py:96]:
```suggestion
    def test_remove_collation_text_type(self):
        remove_collation = SQLConnector.remove_collation
```

File: tests/core/test_connector_sql.py [hunk 7]
```diff
@@ -91,3 +91,42 @@ def test_get_column_ddl(
...
+        test_collation = "SQL_Latin1_General_CP1_CI_AS"
+        current_type = sqlalchemy.types.Text(collation=test_collation)
+        current_type, current_type_collation = remove_collation(current_type)
+        # Check collation was set to None by the function
+        assert current_type.collation == None
+        # Check that we get the same collation we put in back out
+        assert current_type_collation == test_collation
+
+    def test_remove_collation_non_text_type(self, connector):
+        remove_collation = getattr(connector, "remove_collation")
```
Comment (edgarrmondragon) [inline tests/core/test_connector_sql.py:106]:
```suggestion
    def test_remove_collation_non_text_type(self):
        remove_collation = SQLConnector.remove_collation
```

File: tests/core/test_connector_sql.py [hunk 8]
```diff
@@ -91,3 +91,42 @@ def test_get_column_ddl(
...
+        current_type, current_type_collation = remove_collation(current_type)
+        # Check there is not a collation attribute
+        assert hasattr(current_type, "collation") == False
+        # Check that we get the same type we put in
+        assert str(current_type) == "INTEGER"
+        # Check that this variable is missing
+        assert not current_type_collation
+
+    def test_update_collation_text_type(self, connector):
+        update_collation = getattr(connector, "update_collation")
```
Comment (edgarrmondragon) [inline tests/core/test_connector_sql.py:117]:
```suggestion
    def test_update_collation_text_type(self):
        update_collation = SQLConnector.update_collation
```

File: tests/core/test_connector_sql.py [hunk 9]
```diff
@@ -91,3 +91,42 @@ def test_get_column_ddl(
...
+        test_collation = "SQL_Latin1_General_CP1_CI_AS"
+        compatible_type = sqlalchemy.types.Text(collation=None)
+        compatible_type = update_collation(compatible_type, test_collation)
+        # Check collation was set to the value we put in
+        assert compatible_type.collation == test_collation
+
+    def test_update_collation_non_text_type(self, connector):
+        update_collation = getattr(connector, "update_collation")
```
Comment (edgarrmondragon) [inline tests/core/test_connector_sql.py:125]:
```suggestion
    def test_update_collation_non_text_type(self):
        update_collation = SQLConnector.update_collation
```

**Review by edgarrmondragon — CHANGES_REQUESTED** [review]
@BuzzCutNorman `remove_collation` and `update_collation` both mutate the type object _in-place_, i.e. you'll notice the returned value points to the same object as the argument:

```python
new_type = remove_collation(current_type)
assert new_type is current_type
```

So, do you think it makes to only return the collation from `remove_collation` and return nothing from `update_collation`. Similar to operations on a dictionary:

```python
>>> d = {1: 2, 2: 3, 3: 4}
>>> d.pop(1)  # Return the removed value
2
>>> d.update(a=0)  # Return nothing, mutate in-place
>>> d
{2: 3, 3: 4, 'a': 0}
```

**Comment by BuzzCutNorman** [comment]
> So, do you think it makes to only return the collation from `remove_collation` and return nothing from `update_collation`. Similar to operations on a dictionary:

@edgarrmondragon, Yes it does make sense and it looks much cleaner.

**Review by edgarrmondragon — APPROVED** [review]
Thanks @BuzzCutNorman, this is g2g!

**Comment by BuzzCutNorman** [comment]
Thank you @edgarrmondragon ! Couldn't have done it without your excellent guidance.  I learned a lot today.

**Pull request merged**